telemetry/telemetry.go

```diff
@@ -0,0 +1,406 @@
+	bufferItemCount = 0
+	bufferMu.Unlock()
+	return bufCopy
+}
+
+// Response contains the body of a response from the
+// telemetry server.
+type Response struct {
+	// NextUpdate is how long to wait before the next update.
+	NextUpdate time.Duration `json:"next_update"`
+
+	// Stop instructs the telemetry server to stop sending
+	// telemetry. This would only be done under extenuating
+	// circumstances, but we are prepared for it nonetheless.
+	Stop bool `json:"stop,omitempty"`
+
+	// Error will be populated with an error message, if any.
+	// This field should be empty if the status code is < 400.
+	Error string `json:"error,omitempty"`
+
+	// DisableKeys will contain a list of keys/metrics that
+	// should NOT be sent until further notice. The client
+	// must NOT store these items in its buffer or send them
+	// to the telemetry server while they are disabled. If
+	// this list and EnableKeys have the same value (which is
+	// not supposed to happen), this field should dominate.
+	DisableKeys []string `json:"disable_keys,omitempty"`
+
+	// EnableKeys will contain a list of keys/metrics that
+	// MAY be sent until further notice.
+	EnableKeys []string `json:"enable_keys,omitempty"`
+}
+
+// Payload is the data that gets sent to the telemetry server.
+type Payload struct {
+	// The universally unique ID of the instance
+	InstanceID string `json:"instance_id"`
+
+	// The UTC timestamp of the transmission
+	Timestamp time.Time `json:"timestamp"`
+
+	// The timestamp before which the next update is expected
+	// (NOT populated by client - the server fills this in
+	// before it stores the data)
+	ExpectNext time.Time `json:"expect_next,omitempty"`
+
+	// The metrics
+	Data map[string]interface{} `json:"data,omitempty"`
+}
+
+// Int returns the value of the data keyed by key
+// if it is an integer; otherwise it returns 0.
+func (p Payload) Int(key string) int {
+	val, _ := p.Data[key]
+	switch p.Data[key].(type) {
+	case int:
+		return val.(int)
+	case float64: // after JSON-decoding, int becomes float64...
+		return int(val.(float64))
+	}
+	return 0
+}
+
+// countingSet implements a set that counts how many
+// times a key is inserted. It marshals to JSON in a
+// way such that keys are converted to values next
+// to their associated counts.
+type countingSet map[interface{}]int
+
+// MarshalJSON implements the json.Marshaler interface.
+// It converts the set to an array so that the values
+// are JSON object values instead of keys, since keys
+// are difficult to query in databases.
+func (s countingSet) MarshalJSON() ([]byte, error) {
+	type Item struct {
+		Value interface{} `json:"value"`
+		Count int         `json:"count"`
+	}
+	var list []Item
+
+	for k, v := range s {
+		list = append(list, Item{Value: k, Count: v})
+	}
+
+	return json.Marshal(list)
+}
+
+var (
+	// httpClient should be used for HTTP requests. It
+	// is configured with a timeout for reliability.
+	httpClient = http.Client{Timeout: 1 * time.Minute}
```

Comment:
Allow for a custom `tls.Config` (`http.Transport{TLSClientConfig: …}` and so forth) for these reasons:

1. Use *TLS tickets* in your client. ;-)  
```go
myTLSSessionCache = tls.NewLRUClientSessionCache(0) // capacity < 1 → use a default capacity
```
2. Configure a list of known TLS cipher suites.  
Some forks of Go (like the one I use) might have suites (and curves) unknown to your endpoint. By using said list you avoid operators reporting errors for which they cannot completely describe the environment.

Reply:
I'd configure HTTP 1.0 (and no keep-alive).

The *session cache* does not interfere with HTTP/2. And I wouldn't expect it to break TLSv1.3 once that one is in.

Reply:
Well, I was referring more to HTTP/2 needing to be specially configured if creating your own tls.Config on the client-side: `// If non-nil, HTTP/2 support may not be enabled by default.` -- it doesn't really say what is required to enable h2.

But I'm down for no keepalive (the receiving server will enforce this too) -- since I don't really see the benefit of it in this case.

Thanks for your reviews/feedback, Mark!

Reply:
TBH, I wouldn't care if this is HTTP/2 or not. It's one request every so often (1h I gather). Scenarios:

* Use HTTP/2, but go through the whole TLS handshake every time (+how many roundtrips is that‽).
* HTTP 1.0, but resumed sessions. IW on Linux is 10 or higher, giving you at least 12000 octets in one go. Better encoding of HTTP headers could be negligible here, and sth. like head-of-line–blocking is irrelevant when just pushing some JSON.

Reply:
```
The DefaultTransport supports HTTP/2.
// To explicitly enable HTTP/2 on a transport, use golang.org/x/net/http2
// and call ConfigureTransport. See the package docs for more about HTTP/2.
```


https://github.com/golang/go/blob/cd1976dbef0a4a474cccb9e7fcb1e9cf2310df66/src/net/http/transport.go#L72-L74

Reply:
Thanks Eldin. I'll look into that, I guess it's true we don't really need h2 for this.

That's a clever trick -- although I think we have other ways to measure that too.

Reply:
> I guess it's true we don't really need h2 for this.

We don't really need it, but it is nice to have and h2 is the future.